lib/tree.rb

```diff
@@ -2,59 +2,127 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) is worst case if tree is not balanced where n is the height of the tree
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+
+    if @root == nil
+      @root = new_node
+      return
+    end
+
+    current = @root
+
+    while current != nil
+      if current.key >= key
+        break if current.left == nil
```

Comment:
This works, but you could streamline the if statements to insert the node here I think.